# Finding *Rx* in Potato - an example workflow

Example output files are provided in the example_outputs directory. These can be compared to the files generated by this workflow (eg. by checksums) to confirm correct functionality.

## Download required files

All files, except for the reference DM potato genome are available in this repository, simply clone down the repository to your local machine. All commands in this workflow will assume you are based in this repository, you may need to change commands if you change the structure.

Ensure you have followed the instructions in README.md to install required software and optionally set up a cluster profile.

```bash
# Fetch the DM genome, this is too large to include here
cd example_inputs/agrenseq
wget http://spuddb.uga.edu/data/PGSC_DM_v4.03_pseudomolecules.fasta.zip
unzip PGSC_DM_v4.03_pseudomolecules.fasta.zip
```

## Obtain reads

Reads have been submitted to ENA BioProjects PRJEB56823 and PRJEB56825. In this example we will use the GNU parallel utility to download reads. Replace X in the parallel command with the number of jobs to run in parallel, this will depend on your system, based on CPU availability and networking bandwidth. If you prefer another method you may use this, though you may need to change file paths in later files.

```bash
mkdir -p ../Reads
cd ../Reads
parallel -j X wget < ../reads.txt
```

## Perform SMRT-RenSeq assembly

Depending on your system, you may be able to wrap these commands into a job with eg. sbatch. In case you are not able to capture the stdout and stderr from the core snakemake process, log files will be written to .snakemake/log/*.snakemake.log

```bash
# Copy files to SMRT-RenSeq assembly directory
cd ../../smrtrenseq_assembly
cp ../example_inputs/smrtrenseq_assembly/* config/.

# Run workflow
# If using a cluster profile
snakemake --profile /path/to/cluster/profile

# If running locally
snakemake --use-conda --cores=/number/of/CPU/cores
```

## Perform AgRenSeq

```bash
# Copy files to AgRenSeq directory
cd ../agrenseq
cp ../example_inputs/agrenseq/* config/.

# Run workflow
# If using a cluster profile
snakemake --profile /path/to/cluster/profile

# If running locally
snakemake --use-conda --cores=/number/of/CPU/cores
```

## Perform dRenSeq

```bash
# Prepare files from output of SMRT-RenSeq assembly and AgRenSeq

cd ../drenseq

awk '/^>/ {printf("\n%s\n",$0);next; } { printf("%s",$0);}  END {printf("\n");}' < ../smrtrenseq_assembly/assembly/HR02_Gemson/HR02_Gemson.contigs.fasta | tail -n +2 > ../smrtrenseq_assembly/assembly/HR02_Gemson/HR02_Gemson_unwrapped.contigs.fasta # unwrap fasta file so all the sequence is on one line

cat ../smrtrenseq_assembly/assembly/HR02_Gemson/HR02_Gemson_unwrapped.contigs.fasta | grep -A1 -f ../agrenseq/results/HR02_Gemson_filtered_contigs.txt | sed 's/--//g' | sed '/^$/d' > config/HR02_Gemson_candidates.fa # get your sequences for contigs you want

cat ../smrtrenseq_assembly/NLR_Annotator/HR02_Gemson_NLR_Annotator.txt | grep -f ../agrenseq/results/HR02_Gemson_filtered_contigs.txt | less -S # See how many nlrs per contig

# Manually fix file in your favourite text editor - this example uses nano
# Lines can be cut with the shortcut Ctrl + K (copied with Alt + ^ ) and paste
# with Ctrl + U . To cut or copy multiple lines press the shortcut multiple times.
nano config/HR02_Gemson_candidates.fa

cat ../smrtrenseq_assembly/NLR_Annotator/HR02_Gemson_NLR_Annotator.txt | grep -f ../agrenseq/results/HR02_Gemson_filtered_contigs.txt | cut -f2,4-5 > config/HR02_Gemson_candidates.bed # Make a bed file

# Copy files to dRenSeq directory
cp ../example_inputs/drenseq/* config/.

# Run workflow
# If using a cluster profile
snakemake --profile /path/to/cluster/profile

# If running locally
snakemake --use-conda --cores=/number/of/CPU/cores
```

This should show tig00001343_nlr_1 as the only candidate with 100% coverage in all samples scored as 1 in the read scores file used for AgRenSeq. The nucleotide sequence for this can be found in the NLR Annotator output fasta in the SMRT-RenSeq assembly directory. When BLASTed against the NCBI nr/nt database, the top hit is the reference *Rx* sequence.
